File: Linked-List-I/046-reverse-list/reverse-list.py
# class Node:
#   def __init__(self, val):
#     self.val = val
#     self.next = None

# O(n) time and O(n) space
def reverse_list(head, prev = null):
  if (head == None):
    return prev
  next = head.next
  head.next = prev
  return reverse_list(next, head)
  

File: Linked-List-I/046-reverse-list/reverse-list.js
// class Node {
//   constructor(val) {
//     this.val = val;
//     this.next = null;
//   }
// }

// O(n) time and O(1) space
const reverseList = (head) => {
  let prev = null;
  let current = head;

  while (current !== null) {
    let tempNext = current.next;
    current.next = prev;
    prev = current;
    current = tempNext;
  }

  return prev;
};

module.exports = {
  reverseList,
};
